common/include/camera/ICameraControl.hpp

```diff
@@ -0,0 +1,312 @@
+                return glm::epsilonEqual(glm::dot(a, b), 0.0f, epsilon);
+            };
+
+            auto isOrthoBase = [&](const auto& x, const auto& y, const auto& z, float epsilon = 1e-6f) -> bool
+            {
+                return isNormalized(x, epsilon) && isNormalized(y, epsilon) && isNormalized(z, epsilon) &&
+                    isOrthogonal(x, y, epsilon) && isOrthogonal(x, z, epsilon) && isOrthogonal(y, z, epsilon);
+            };
+
+            assert(isOrthoBase(xaxis, yaxis, zaxis));
+        }
+
+        float32_t3 m_position;
+        glm::quat m_orientation;
+
+        // Represents the camera's orthonormal basis
+        // https://en.wikipedia.org/wiki/Orthonormal_basis
+        float32_t3x3 m_orthonormal;
+    };
+
+    ICameraController(core::smart_refctd_ptr<CGimbal>&& gimbal) : m_gimbal(core::smart_refctd_ptr(gimbal)) {}
+
+    // override controller keys map, it binds a key code to a virtual event
+    void updateKeysToEvent(const std::vector<ui::E_KEY_CODE>& codes, VirtualEventType event)
+    {
+        m_keysToEvent[event] = std::move(codes);
+    }
+
+    // start controller manipulation session
+    virtual void begin(std::chrono::microseconds nextPresentationTimeStamp)
+    {
+        m_nextPresentationTimeStamp = nextPresentationTimeStamp;
+        return;
+    }
+
+    // manipulate camera with gimbal & virtual events, begin must be called before that!
+    virtual void manipulate(std::span<const CVirtualEvent> virtualEvents) = 0;
+
+    // finish controller manipulation session, call after last manipulate in the hot loop
+    void end(std::chrono::microseconds nextPresentationTimeStamp)
+    {
+        m_lastVirtualUpTimeStamp = nextPresentationTimeStamp;
+    }
+
+    /*
+    // process keyboard to generate virtual manipulation events
+    // note that:
+    // - it doesn't make the manipulation itself!
+    */
+    std::vector<CVirtualEvent> processKeyboard(std::span<const ui::SKeyboardEvent> events)
+    {
+        if (events.empty())
+            return {};
+
+        std::vector<CVirtualEvent> output;
+
+        constexpr auto NblVirtualKeys = std::to_array({ MoveForward, MoveBackward, MoveLeft, MoveRight, MoveUp, MoveDown, TiltUp, TiltDown, PanLeft, PanRight, RollLeft, RollRight, Reset });
+        static_assert(NblVirtualKeys.size() == EventsCount);
+
+        for (const auto virtualKey : NblVirtualKeys)
+        {
+            const auto code = m_keysToEvent[virtualKey];
+            bool& keyDown = m_keysDown[virtualKey];
+
+            using virtual_key_state_t = std::tuple<ui::E_KEY_CODE /*physical key representing virtual key*/, bool /*is pressed*/, float64_t /*delta action*/>;
+
+            auto updateVirtualState = [&]() -> virtual_key_state_t
+            {
+                virtual_key_state_t state = { ui::E_KEY_CODE::EKC_NONE, false, 0.f };
+
+                for (const auto& ev : events) // TODO: improve the search
+                {
+                    if (ev.keyCode == code)
```

Comment:
maintain a simple table of all key codes to actions/virtual keys (and maybe even callbacks/lambdas to execute on them)

Reply:
resolved